### DISCUSSION TITLE
PluginAPI to capture mouse response

### DISCUSSION THREAD

**Discussion by Hctor99** [jspsych/jsPsych#992] [Q&A] [answered]
Hello all!

I'm working on a customized plugin which will allow me to get the x,y coordinates from the place on the screen where the user clicked on. I was wondering if there was a PluginAPI such as jsPsych.pluginAPI.getKeyboardResponse that would wait for a mouse click response (or will getKeyboardResponse do the trick somehow?). If that is not the case, do you have any ideas of how I can do this?

If this helps, here's what I'm doing.

I found on the internet some html code which has the following lines in the head:

```
<script>
    document.onclick = function(e) { 
      coords.innerHTML = e.clientX + ':' + e.clientY;
    };
 </script>
```

And the following line in the body:

`<div id="coords">(click coordinates show up here)</div>
`

The coordinates of the mouse click show up in the last line, as it is indicated. I just want to get that data after the user has clicked on the screen.

Thank you!

**Comment by jodeleeuw** [comment, marked as answer]
Hi @Hctor99 ,

There's no function in the `pluginAPI` module that will do this, but I think you're pretty close with the code you've found so far.

```js
document.onclick = function(e) { 
      coords.innerHTML = e.clientX + ':' + e.clientY;
};
```

This part of the code sets up an event listener. It runs the function every time the document is clicked. The function modifies the HTML inside the `<div>` tag. `e.clientX` and `e.clientY` are the X/Y coodinates of the click event. Note that `e` matches the parameter name inside the declaration of the function.

To modify this to work for a plugin, you could capture the location into a variable instead.

```js
var click_x;
var click_y;
document.onclick = function(e) { 
      	click_x = e.clientX;
	click_y = e.clientY;
};
```

Then you would need to add this to the data for the plugin. The [documentation on creating a new plugin](https://www.jspsych.org/plugins/creating-a-plugin/) describes how this works. I also recorded [a video tutorial about building new plugins](https://www.youtube.com/watch?v=XQcsFwAmbiw) that you may find helpful.

Finally, one thing that you may want to do here is use the function `addEventListener()` instead of using the `.onclick` property of the `document`. This is because you'll want to *remove* the event listener at the end of the plugin.

```js
var click_x;
var click_y;
var click_listener = function(e) { 
      	click_x = e.clientX;
	click_y = e.clientY;
});
document.addEventListener('click', click_listener);

// at the end of the plugin
document.removeEventListener('click', click_listener);
```

**Reply by Hctor99** [reply]
Hey @jodeleeuw, thank you so much for your detailed response! I finally got to make it work :)

**Comment by kurokida** [comment]
Although this issue was solved, I introduce a plugin.
The jspsych-psychophysics plugin:
http://jspsychophysics.hes.kyushu-u.ac.jp/
